python/ql/src/semmle/python/functions/ModificationOfParameterWithDefaultCustomizations.qll

```diff
@@ -0,0 +1,217 @@
+      or
+      // modifying function call
+      exists(DataFlow::CallCfgNode c, DataFlow::AttrRead a | c.getFunction() = a |
+        a.getObject() = this and
+        a.getAttributeName() in [list_modifying_method(), dict_modifying_method()]
+      )
+    }
+  }
+
+  /**
+   * An expression that is checked directly in an `if`, possibly with `not`, such as `if x:` or `if not x:`.
+   */
+  private class IdentityGuarded extends Expr {
+    boolean inverted;
+
+    IdentityGuarded() {
+      this = any(If i).getTest() and
+      inverted = false
+      or
+      exists(IdentityGuarded ig, UnaryExpr notExp |
+        notExp.getOp() instanceof Not and
+        ig = notExp and
+        notExp.getOperand() = this
+      |
+        inverted = ig.isInverted().booleanNot()
+      )
+    }
+
+    /**
+     * Whether this guard has been inverted. For `if x:` the result is `false`, and for `if not x:` the result is `true`.
+     */
+    boolean isInverted() { result = inverted }
+  }
+
+  /**
+   * Holds iff `guard` is checking the `Name` represented by `guarded` for truthyness.
+   * `result` is true if the check is inverted and false if it is not.
+   */
+  boolean isIdentityGuard(DataFlow::GuardNode guard, ControlFlowNode guarded) {
+    exists(IdentityGuarded ig |
+      ig instanceof Name and
+      // In `not l`, the `ControlFlowNode` for `l` is not an instance of `GuardNode`.
+      // TODO: This is slightly naive, not handling e.g. `l or cond` correctly.
+      // We should change it when we have a proper guards library.
+      guard.getNode().getAChildNode*() = ig and
+      result = ig.isInverted() and
+      guarded.getNode() = ig
+    )
+  }
+
+  /**
+   * A sanitizer guard that does not let a truthy value flow to the true branch.
+   * Based on `isIdentityGuard`, so comes with the same caveats.
+   */
+  class BlocksTruthyGuard extends BlocksTruthy {
+    ControlFlowNode guarded;
+
+    BlocksTruthyGuard() {
+      // The raw guard is true if the value is non-empty.
+      // We wish to send truthy falues to the false branch,
+      // se we are looking for inverted guards.
+      isIdentityGuard(this, guarded) = true
+    }
+
+    override predicate checks(ControlFlowNode node, boolean branch) {
+      node = guarded and
+      branch = true
+    }
+  }
+
+  /**
+   * A sanitizer guard that does not let a falsy value flow to the true branch.
+   * Based on `isIdentityGuard`, so comes with the same caveats.
+   */
+  class BlocksFalseyGuard extends BlocksFalsey {
+    ControlFlowNode guarded;
+
+    BlocksFalseyGuard() {
+      // The raw guard is true if the value is non-empty.
+      // We wish to send falsy falues to the false branch,
+      // se we are looking for guards that are not inverted.
+      isIdentityGuard(this, guarded) = false
+    }
+
+    override predicate checks(ControlFlowNode node, boolean branch) {
+      node = guarded and
+      branch = true
+    }
+  }
```

Comment:
As discussed offline

```suggestion
  /**
   * A sanitizer guard that does not let a truthy value flow to the true branch.
   *
   * Blocks flow of `x` in the true branch in the example below.
   * ```py
   * if x:
   *     x.append(42)
   * ```
   * 
   * Implementation note: Since guards with different behaviour cannot exist on the same
   * node, we let all guards have the same behaviour, in the sense that they all check
   * the true branch. Instead, we partition guards into those that block truthy values
   * and those that block falsy values.
   */
  class BlocksTruthy extends DataFlow::BarrierGuard {
    BlocksTruthy() {
      this instanceof NameNode
    }

    override predicate checks(ControlFlowNode node, boolean branch) {
      node = this and
      branch = true
    }
  }

  /**
   * A sanitizer guard that does not let a falsey value flow to the true branch.
   *
   * Blocks flow of `x` in the true branch in the example below.
   * ```py
   * if not x:
   *     x.append(42)
   * ```
   * 
   * Implementation note: Since guards with different behaviour cannot exist on the same
   * node, we let all guards have the same behaviour, in the sense that they all check
   * the true branch. Instead, we partition guards into those that block truthy values
   * and those that block falsy values.
   */
  class BlocksFalsey extends DataFlow::BarrierGuard {
    NameNode guarded;

    BlocksFalsey() {
      guarded = this.(UnaryExprNode).getOperand() and
      this.(UnaryExprNode).getNode().getOp() instanceof Not
    }

    override predicate checks(ControlFlowNode node, boolean branch) {
      node = guarded and
      branch = true
    }
  }
```